From fbbe0aba17e13d951ce919a7b41f26dadd5a21f1 Mon Sep 17 00:00:00 2001
From: Fancy Fang <chen.fang@nxp.com>
Date: Fri, 25 Aug 2017 17:30:53 +0800
Subject: [PATCH 2447/5242] MLK-16255-5 video: fbdev: dcss: init default video
 layer buffer to black

commit  0b4f4f0420fa2188f0d65ad4ee5e8a991b9f8cd6 from
https://source.codeaurora.org/external/imx/linux-imx.git

For video layer, the black pixel value is not 0x0,
but (Y: 0x0, U: 128, V: 128). So init the video
layer frame buffer contents to the black pixel
value by default.

Signed-off-by: Fancy Fang <chen.fang@nxp.com>
Signed-off-by: Meng Li <Meng.Li@windriver.com>
---
 drivers/video/fbdev/mxc/imx_dcss.c |   19 ++++++++++++++++---
 1 file changed, 16 insertions(+), 3 deletions(-)

diff --git a/drivers/video/fbdev/mxc/imx_dcss.c b/drivers/video/fbdev/mxc/imx_dcss.c
index 0a2f185..8ae9186 100644
--- a/drivers/video/fbdev/mxc/imx_dcss.c
+++ b/drivers/video/fbdev/mxc/imx_dcss.c
@@ -995,6 +995,7 @@ static int dcss_init_chans(struct dcss_info *info)
 static int dcss_init_fbinfo(struct fb_info *fbi)
 {
 	int ret;
+	uint32_t luma_size;
 	struct dcss_channel_info *cinfo = fbi->par;
 	struct dcss_info *info = cinfo->dev_data;
 	struct fb_fix_screeninfo *fix = &fbi->fix;
@@ -1054,9 +1055,21 @@ static int dcss_init_fbinfo(struct fb_info *fbi)
 		return -ENOMEM;
 	}
 
-
-	/* clear screen content */
-	memset((void*)fbi->screen_base, 0x0, fix->smem_len);
+	/* clear screen content to black */
+	switch (var->grayscale) {
+	case V4L2_PIX_FMT_ARGB32:
+		memset((void*)fbi->screen_base, 0x0, fix->smem_len);
+		break;
+	case V4L2_PIX_FMT_NV12:
+		/* set luma: 0x0 */
+		luma_size = var->xres * var->yres;
+		memset((void*)fbi->screen_base, 0x0, luma_size);
+		/* set chroma: 0x80 */
+		memset((void*)fbi->screen_base + luma_size, 0x80, luma_size);
+		break;
+	default:
+		return -EINVAL;
+	}
 
 	if (dcss_check_var(var, fbi)) {
 		devm_kfree(fbi->device, fbi->pseudo_palette);
-- 
1.7.9.5

